cli_tools/common/utils/param/param_utils.go

```diff
@@ -0,0 +1,127 @@
+//  Copyright 2019 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+//+build !test
+
+package paramutils
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/common/domain"
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/common/utils/storage"
+	"github.com/GoogleCloudPlatform/compute-image-tools/daisy/compute"
+	"google.golang.org/api/option"
+)
+
+// Get project id from either current project or flag
+func GetProjectID(mgce commondomain.MetadataGCEInterface, projectFlag string) (string, error) {
+	if projectFlag == "" {
+		if !mgce.OnGCE() {
+			return "", fmt.Errorf("project cannot be determined because build is not running on GCE")
+		}
+		aProject, err := mgce.ProjectID()
+		if err != nil || aProject == "" {
+			return "", fmt.Errorf("project cannot be determined %v", err)
+		}
+		return aProject, nil
+	}
+	return projectFlag, nil
+}
+
+
+func PopulateMissingParameters(project *string, zone *string, region *string, scratchBucketGcsPath *string, file string, mgce commondomain.MetadataGCEInterface,
+		scratchBucketCreator commondomain.ScratchBucketCreatorInterface,
+		zoneRetriever commondomain.ZoneRetrieverInterface, storageClient commondomain.StorageClientInterface) error {
+
+	if err := PopulateProjectIfMissing(mgce, project); err != nil {
+		return err
+	}
+
+	scratchBucketRegion := ""
+	if *scratchBucketGcsPath == "" {
+		scratchBucketName, sbr, err := scratchBucketCreator.CreateScratchBucket(file, *project)
+		scratchBucketRegion = sbr
+		if err != nil {
+			return err
+		}
+
+		*scratchBucketGcsPath = fmt.Sprintf("gs://%v/", scratchBucketName)
+	} else {
+		scratchBucketName, err := storageutils.GetBucketNameFromGCSPath(*scratchBucketGcsPath)
+		if err != nil {
+			return fmt.Errorf("invalid scratch bucket GCS path %v", *scratchBucketGcsPath)
+		}
+		scratchBucketAttrs, err := storageClient.GetBucketAttrs(scratchBucketName)
+		if err == nil {
+			scratchBucketRegion = scratchBucketAttrs.Location
+		}
+	}
+
+	if *zone == "" {
+		if aZone, err := zoneRetriever.GetZone(scratchBucketRegion, *project); err == nil {
+			*zone = aZone
+		} else {
+			return err
+		}
+	}
+
+	if err := PopulateRegion(region, *zone); err != nil {
+		return err
+	}
+	return nil
+}
+
+func PopulateProjectIfMissing(mgce commondomain.MetadataGCEInterface, projectFlag *string) error {
```

Comment:
I don't quite agree, the details should be hidden, this function is an abstraction. Details are explained by GetProjectID.

Reply:
Since PopulateProjectIfMissing has GCE metadata as an argument, documenting how it's used is appropriate. Otherwise next engineer using this function will wonder how it's used, read the source code and figure it out for themselves.

Reply:
My opinion: Next engineer will found the details from comment of GetProjectID. The real purpose of PopulateProjectIfMissing is to populate whatever project id decided by GetProjectID. For this layer, the answer of "how it's used" is "pass to GetProjectID"
Let me know whether that makes sense :D